### DISCUSSION TITLE
Buttons and Events (like onClick) work in development but break in production

### DISCUSSION THREAD

**Discussion by rons613** [vercel/next.js#16032] [Help] [answered]
# Bug report

Events are not being detected (and therefore buttons don't work) in Next JS production builds. However, everything works fine in development.

## Describe the bug

Browser events (like onClick) are non-existent in production versions of app. No errors are thrown when compiling and building production app using "next build".  Switching my current version of next with next@canary as described in [issue #6568](https://github.com/vercel/next.js/issues/6588) did not solve the problem. 

## To Reproduce

1. Create a simple Next app similar to the one in [this repository](https://github.com/rons613/broken_button_nextjs).

2. Build next app using npm run build. 

3. Start production app using NODE_ENV=production node server.js

4. Try clicking on button to trigger a state change from "some state" to "another state". 

5. In production, no state change occurs nor are any events visible in browser console. 

## Expected behavior

Button onClick event is supposed to trigger state change in both development and production. However, nothing happens in production when button is clicked.

## System information

- OS: MacOS High Sierra
- Browser (if applies) Chrome, Firefox, Safari
- Version of Next.js: 9.4.4
- Version of Node.js: 12.6.1 in production, 11.10.0 in development

## Additional context

In development environment, state change occurs every time button is clicked. Also, the console.log message is displayed each time in browser's development console. Event listeners are visible in browser for development version of app. However, nothing works in production. No state change occurs when button is clicked, no logs are displayed and event listeners are non-existent in browser consoles.

**Comment by rons613** [comment]
I see that this bug report has been converted into a discussion. Are there any solutions to this issue besides using next@canary (which was mentioned in a similar issue but isn't working in my case)? If so, can anyone make suggestions even if you think this report is trivial. I'd hate to have to scrap my project (larger in scope than the simplified repo example) and switch to a different tech stack because of this bug in production. I like the benefits of Next JS (and would like to stick to it in the long run) but can't afford to have such issues with production deployments going unresolved.

**Comment by mattcarlotta** [comment, marked as answer]
This is unrelated to Next. There are some code choices that should be fixed which should fix your problem.

For example, you're invoking `event.preventDefault` within the `changeSomething` class property; however, this is not needed as it's not preventing any default event action (if button was type `submit` and it's a child of a form, then `event.preventDefault` would prevent the page from refreshing -- in this case, since its type is `button`, then it prevents nothing). Also you're also passing in `this.state.someState` (`oldState`) as an argument to `changeSomething` when the state is already available via `this.state`.

You can simplify your code to the following example below (confirmed working in development and production -- comments are an optional way of using traditional functions (`function(){ ... })` as class properties where `this` needs to be manually `bind`ed to the class property, otherwise `this` will be undefined. When fat arrow functions (`() => { ... }`) are used as class properties, then  they retain the lexical scope of `this` and don't require `this` to be manually `bind`ed to it. As to which one to use, it's a personal preference):
```js
import * as React from "react";
import Head from "next/head";

class HomePage extends React.Component {
  // constructor(props) {
  //   super(props);

  //   this.state = {
  //     someState: "some state",
  //   };

  //   this.changeSomething = this.changeSomething.bind(this);
  // }

  static async getInitialProps() {
    return { someProp: "a random prop" };
  }

  state = {
    someState: "some state",
  };

  // changeSomething() {
  //   console.log("Button was clicked");
  //   const { someState } = this.state;

  //   let newState;
  //   if (someState == "some state") {
  //     newState = "another state";
  //   } else {
  //     newState = "some state";
  //   }

  //   this.setState({ someState: newState });
  // };

  changeSomething = () => {
    console.log("Button was clicked");
    const { someState } = this.state;

    let newState;
    if (someState == "some state") {
      newState = "another state";
    } else {
      newState = "some state";
    }

    this.setState({ someState: newState });
  };

  render() {
    return (
      <React.Fragment>
        <Head>
          <title>The Home Page</title>
        </Head>
        <p>Home Page State: {this.state.someState}</p>
        <p>Initial Prop: {this.props.someProp}</p>
        <div>
          <button type="button" onClick={this.changeSomething}>
            CHANGE STATE PARAMETER
          </button>
        </div>
      </React.Fragment>
    );
  }
}

export default HomePage;

```

**Reply by rons613** [reply]
Thanks for the help @mattcarlotta . But it's still not working in production for me even when I replace my home page with your example. Are you using the same dependencies (including versions of them). In my production setup, I have this app running behind Nginx on a cloud-hosted server. Could that be a part of the problem?

Edit: Actually, when I run it in production mode on my laptop and go to localhost:3000, it works. So this must be an issue with hosting it behind Nginx on my VPS (referenced by a unique domain name). I'll have to go troubleshoot that. Thanks for all your help. I appreciate it.

**Reply by mattcarlotta** [reply]
I copied your repo, made the adjustments above, and ran a local development & production build. Both worked fine. I'd recommend eliminating as many variables as possible by running **just** the example repo in both local development/production and in remote production and seeing if you can achieve a working solution. Without seeing any code (aside from the example repo you've provided), my guess would be that the Javascript isn't being served/loaded in your remote production.

**Reply by rons613** [reply]
I ran the repo app in production on my local development/production environment and on the remote VPS. It works on my local machine but fails on the server. My original app also has the same behavior (working locally but not remotely). Up until now, I always built production apps on the server (but never locally) since I trusted that they'd work on the VPS (like another Next JS app I had previously written a while back which worked fine). 

But you are right that it's an issue with Javascript not being loaded. In Firefox's developer console, there are even messages which say, "Loading failed for the <script> with source ..........(name of source file)". It's far from a Next JS issue which I'll continue  trying to solve. But I am feeling more confident about solving all my issues eventually thanks to your feedback.

**Reply by yuhcee** [reply]
This is the same behavior for me currently @rons613. Where you able to fix it later on? Sometimes Chrome does not load up the script of my register component in nextjs and thus my submit button does not work, however, sometimes it will just work fine. I don't seem to understand what is wrong with it. Care to share how you were finally able to resolve this, please? Thanks.

**Reply by rons613** [reply]
@uchennaegbo I wrote another post mentioning where it went wrong for me. It was due to issues with my Nginx reverse proxy as mentioned here:

https://github.com/vercel/next.js/discussions/16032#discussioncomment-49347

**Comment by rons613** [comment]
Thanks to the greatly appreciated feedback of @mattcarlotta , it was determined that this was not a Next JS issue. I also managed to get my code repo and original app working without any problems in production (both locally and on my cloud-hosted virtual private server).  

Just in case anyone runs into the same problem, I'll post where I went wrong. In local development and production, the app worked because all I had to do was go to localhost:3000 to see the app. However, on my remote server (containing a domain name), I had my app running on localhost behind Nginx as a reverse proxy. I had Nginx set up in such a way that it would only forward requests to my Next JS app if the relative path began with some random string known only to me. For example, mydomainDOTcom/secretstring (secretstring = home in the repo example) would point to the Next JS app while mydomainDOTcom/ (slash nothing), would point to a completely different resource. 

It turns out that when a request was made to the server, it would fetch the main HTML page from the server (with Nginx forwarding the request to the right app thanks to the "/secretstring" rule in my Nginx setup). However, when fetching the necessary Javascript from the server, the request was not forwarded to my Next JS app. Instead it was forwarded to the resource behind the rule for the "/" path. Since that resource doesn't contain the necessary Javascript, no Javascript was loaded and therefore none of my website's buttons worked at all. 

I fixed it by modifying the routing rules directly inside my Next JS app and changed the Nginx setup so that all requests beginning with "/" are forwarded to my Next JS app. Now requests for the HTML and Javascript (which occur separately) are routed to the same app on the server. I wanted to hide my webpages behind a secret relative path so that search engines wouldn't crawl my website before I was ready for them to do so. I just went about it the wrong way when configuring Nginx.

All-in-all, this was never a problem with NextJS but with my server setup. Thanks to @mattcarlotta , I was able to resolve this issue fairly quickly and get over fears about possibly having to use some other frontend framework (which wouldn't have been very desirable given the user-friendliness of Next).

**Comment by Ehab97** [comment]
Actually i have the same problem but i don't have the unneeded syntax , and the problem appear only in root page pages/index.js
do you have any recommendation thanks

**Reply by rons613** [reply]
What worked for me was to just put the entire app behind the "/" route on Nginx. Trying to do this in any other way stops the Javascript from being pulled from the server, since the HTML and Javascript are apparently both loaded separately. 

How are you hosting this app? Is it behind an Nginx server block for a given domain? Are you using Nginx to host other apps as well on the same server?

**Reply by sleemkeen** [reply]
do you have a sample Nginx config for this ????

**Reply by rons613** [reply]
> do you have a sample Nginx config for this ????

In the "/etc/nginx/sites-available/example.com" file, it's best to set it up this way:

server {
...
    location / {
        proxy_pass http://localhost:3000; ###YOUR NEXT.JS APP BEHIND www.example.com
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
    }
...
   location /app2 {
        proxy_pass http://localhost:3001; ##### SOME OTHER NODE APP HIDDEN BEHIND www.example.com/app2 route
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
    }
}

I was initially hosting my Next JS app behind the "/app2". That had the effect of my browser downloading only the HTML from the correct app behind /app2, but then trying to get the Javascript from the "/" route. So I put my NextJS app to the "/" route. Now requests for both the HTML and Javascript go to the same place.
